### ISSUE TITLE
Step time cost is very different between images on FasterRCNN example

### ISSUE THREAD

**Issue by yogurfrul** [tensorpack/tensorpack#858]
When I run FasterRCNN example using HrovodTrainer, 
I find that rank step time cost are very different because of different image shape and number of obeject to detect. So the Negotiation phase in the horovod timeline are very long.

I want to konw that if this means the most time cost are in backbone which has many Conv2D 
and How would I improve the train speed 

When I export TF_CUDNN_USE_AUTOTUNE=0 as you said in  the  [link](https://github.com/tensorflow/tensorflow/issues/13114#issuecomment-330381682) ,the step cost is still very different and no improve of performence

**Comment by ppwwyyxx** [comment]
> if this means the most time cost are in backbone which has many Conv2D

I don't know if this is true. You'll need more detailed benchmarks.

>  How would I improve the train speed

You can let the dataloader produce images of more similar sizes together.

> When I export TF_CUDNN_USE_AUTOTUNE=0 as you said in the link ,the step cost is still very different and no improve of performence

The image sizes are still different so this option does nothing useful for this issue.

**Comment by yogurfrul** [comment]
I want to show op output shape  so I 

```
image_shape2d = tf.shape(image,name='imageshape')[2:] 
```
and I add monitors in TrainConfig
```
monitors=[ScalarPrinter(enable_step=True,whitelist=['.*/Conv2D','imageshape'])],
```
But I get noting to print

**Comment by ppwwyyxx** [comment]
monitors only take care of where to put logging data. It does not manage what to log and when they are produced. See http://tensorpack.readthedocs.io/tutorial/summary.html#tensorflow-summaries

For your use case it's better to read http://tensorpack.readthedocs.io/tutorial/faq.html#how-to-print-dump-intermediate-results-in-training

**Comment by yogurfrul** [comment]
Thx
and I get wrong result after TF_CUDNN_USE_AUTOTUNE=0
I set shuffle=False in get_train_dataflow ,then print the image shape and obeject number in the image and step time cost
then I get the log below:

unset TF_CUDNN_USE_AUTOTUNE
```
mage shape:[1201 800 3]
obeject number:[2]
[32m[0810 06:27:14 @base.py:254][0m hvd rank 0, (global_step 86) finished at 1533882434.3568213, time:0.2963137626647949.
image shape:[800 1199 3]
obeject number:[8]
[32m[0810 06:27:14 @base.py:254][0m hvd rank 0, (global_step 87) finished at 1533882434.6940699, time:0.33693432807922363.
image shape:[800 1261 3]
obeject number:[10]
[32m[0810 06:27:17 @base.py:254][0m hvd rank 0, (global_step 88) finished at 1533882437.2974875, time:2.603098154067993.
image shape:[800 808 3]
obeject number:[2]
[32m[0810 06:27:17 @base.py:254][0m hvd rank 0, (global_step 89) finished at 1533882437.885671, time:0.5877795219421387.
image shape:[800 1278 3]obeject number:[4]

[32m[0810 06:27:20 @base.py:254][0m hvd rank 0, (global_step 90) finished at 1533882440.3424397, time:2.456442356109619.
image shape:[800 1199 3]
obeject number:[1]
[32m[0810 06:27:22 @base.py:254][0m hvd rank 0, (global_step 91) finished at 1533882442.4944735, time:2.151698350906372.
```
export TF_CUDNN_USE_AUTOTUNE=0

```
image shape:[1201 800 3]
obeject number:[2]
[32m[0810 06:16:50 @base.py:254][0m hvd rank 0, (global_step 86) finished at 1533881810.8056705, time:0.379549503326416.
image shape:[800 1199 3]obeject number:[8]

[32m[0810 06:16:51 @base.py:254][0m hvd rank 0, (global_step 87) finished at 1533881811.2121527, time:0.40617823600769043.
image shape:[800 1261 3]
obeject number:[10]
[32m[0810 06:16:51 @base.py:254][0m hvd rank 0, (global_step 88) finished at 1533881811.572611, time:0.36014246940612793.
image shape:[800 808 3]obeject number:[2]

[32m[0810 06:16:52 @base.py:254][0m hvd rank 0, (global_step 89) finished at 1533881812.0219898, time:0.44893598556518555.
image shape:[800 1278 3]
obeject number:[4]
[32m[0810 06:16:52 @base.py:254][0m hvd rank 0, (global_step 90) finished at 1533881812.3514462, time:0.32912302017211914.
image shape:[800 1199 3]
obeject number:[1]
[32m[0810 06:16:52 @base.py:254][0m hvd rank 0, (global_step 91) finished at 1533881812.7503793, time:0.3986361026763916.
```
I just try one epoch of 210 step ,
the epoch cost time are 4min 54s(enable autotune) and 1min 44s(disable autotune)

**Comment by ppwwyyxx** [comment]
What is your question?

**Comment by yogurfrul** [comment]
If each data has different size,  , export TF_CUDNN_USE_AUTOTUNE=0 may improve train performence
It would be very nice to let others know that  disable cudnn autotune in  FasterRCNN  example

**Comment by ppwwyyxx** [comment]
It will improve performance at the beginning as mentioned in the fastrcnn notes.
I don't know whether it improves overall performance or not.

**Comment by yogurfrul** [comment]
I  can only find this in fastrcnn notes
```
The training will start very slow due to convolution warmup, until about 10k steps to reach a maximum speed. Then the training speed will slowly decrease due to more accurate proposals.
```

1,.You mean AUTOTUNE mainly do the convolution warmup?
2.why 10K would reach a maximum speed , there are about 10k kinds of image shape in the COCO?
3.If I want to check the performence of trainning by step cost time and performence  of gradient allreduce , do you have some suggestion?

**Comment by ppwwyyxx** [comment]
1. warmup mainly means autotune
2. the number is a rough estimate and has no mathematical meaning
3. (1) you can write a callback and put a timer in `trigger_step`
    (2) I don't know how to get time spent of a specific op in tensorflow except to enable tracing and parse the result

**Comment by ppwwyyxx** [comment]
What is your question?
Your log shows it's not 0.5it/s but >1.3it/s.

**Comment by yogurfrul** [comment]
my question is  2 gpu  run faster  than 1 gpu without gradient allreduce
I can not find some way to debug it.
CUDA_VISIBLE_DEVICES=0,1 python train.py ,I run many times and speed up from about 60 step ,then come to 4 it/s.
```
  (global_step 137) finished at 1534294900.3835046, time:0.11500763893127441.
[0815 01:01:40 @base.py:247]  (global_step 138) begin at 1534294900.3839097
 66%|###############################################################                                 |138/210[01:33<00:15, 4.53it/s][0815 01:01:40 @base.py:254]  (global_step 138) finished at 1534294900.504368, time:0.12045836448669434.
[0815 01:01:40 @base.py:247]  (global_step 139) begin at 1534294900.5047371
[0815 01:01:40 @base.py:254]  (global_step 139) finished at 1534294900.6229901, time:0.11825299263000488.
[0815 01:01:40 @base.py:247]  (global_step 140) begin at 1534294900.6234102
[0815 01:01:40 @base.py:254]  (global_step 140) finished at 1534294900.7403178, time:0.11690759658813477.
[0815 01:01:40 @base.py:247]  (global_step 141) begin at 1534294900.7407358
[0815 01:01:40 @base.py:254]  (global_step 141) finished at 1534294900.8584886, time:0.1177527904510498.
[0815 01:01:40 @base.py:247]  (global_step 142) begin at 1534294900.8589025
 68%|################################################################9                               |142/210[01:34<00:16, 4.12it/s][0815 01:01:41 @base.py:254]  (global_step 142) finished at 1534294901.5618732, time:0.7029707431793213.
[0815 01:01:41 @base.py:247]  (global_step 143) begin at 1534294901.5622423
[0815 01:01:41 @base.py:254]  (global_step 143) finished at 1534294901.6763937, time:0.1141514778137207.
[0815 01:01:41 @base.py:247]  (global_step 144) begin at 1534294901.676972
[0815 01:01:42 @base.py:254]  (global_step 144) finished at 1534294902.0140057, time:0.337033748626709.
[0815 01:01:42 @base.py:247]  (global_step 145) begin at 1534294902.0145013
 69%|##################################################################2                             |145/210[01:34<00:14, 4.58it/s][0815 01:01:42 @base.py:254]  (global_step 145) finished at 1534294902.143973, time:0.1294717788696289.
[0815 01:01:42 @base.py:247]  (global_step 146) begin at 1534294902.1443706
[0815 01:01:42 @base.py:254]  (global_step 146) finished at 1534294902.8292904, time:0.6849198341369629.
[0815 01:01:42 @base.py:247]  (global_step 147) begin at 1534294902.829791
[0815 01:01:43 @base.py:254]  (global_step 147) finished at 1534294903.529799, time:0.7000079154968262.
[0815 01:01:43 @base.py:247]  (global_step 148) begin at 1534294903.5302541
 70%|###################################################################6                            |148/210[01:36<00:22, 2.78it/s][0815 01:01:43 @base.py:254]  (global_step 148) finished at 1534294903.6478417, time:0.11758756637573242.
[0815 01:01:43 @base.py:247]  (global_step 149) begin at 1534294903.6482408
[0815 01:01:43 @base.py:254]  (global_step 149) finished at 1534294903.772947, time:0.12470626831054688.
[0815 01:01:43 @base.py:247]  (global_step 150) begin at 1534294903.7733433
[0815 01:01:43 @base.py:254]  (global_step 150) finished at 1534294903.88845, time:0.11510658264160156.
[0815 01:01:43 @base.py:247]  (global_step 151) begin at 1534294903.8888514
 72%|#####################################################################                           |151/210[01:36<00:15, 3.75it/s][0815 01:01:44 @base.py:254]  (global_step 151) finished at 1534294904.1664782, time:0.2
```

CUDA_VISIBLE_DEVICES=0 python train.py 

```
 66%|###############################################################5                                |139/210[03:49<01:43, 0.69it/s][0815 01:09:38 @base.py:254]  (global_step 139) finished at 1534295378.650176, time:2.45257830619812.
[0815 01:09:38 @base.py:247]  (global_step 140) begin at 1534295378.650516
 67%|################################################################                                |140/210[03:51<02:08, 0.55it/s][0815 01:09:40 @base.py:254]  (global_step 140) finished at 1534295380.852188, time:2.201672077178955.
[0815 01:09:40 @base.py:247]  (global_step 141) begin at 1534295380.8525262
[0815 01:09:41 @base.py:254]  (global_step 141) finished at 1534295381.1603434, time:0.3078172206878662.
[0815 01:09:41 @base.py:247]  (global_step 142) begin at 1534295381.1607199
 68%|################################################################9                               |142/210[03:54<01:46, 0.64it/s][0815 01:09:43 @base.py:254]  (global_step 142) finished at 1534295383.447481, time:2.2867610454559326.
[0815 01:09:43 @base.py:247]  (global_step 143) begin at 1534295383.447814
[0815 01:09:43 @base.py:254]  (global_step 143) finished at 1534295383.7949831, time:0.3471691608428955.
[0815 01:09:43 @base.py:247]  (global_step 144) begin at 1534295383.7953558
 69%|#################################################################8                              |144/210[03:55<01:02, 1.05it/s][0815 01:09:44 @base.py:254]  (global_step 144) finished at 1534295384.1309144, time:0.3355586528778076.
[0815 01:09:44 @base.py:247]  (global_step 145) begin at 1534295384.1312623
[0815 01:09:44 @base.py:254]  (global_step 145) finished at 1534295384.4696224, time:0.338360071182251.
[0815 01:09:44 @base.py:247]  (global_step 146) begin at 1534295384.4699774
 70%|##################################################################7                             |146/210[03:57<01:08, 0.93it/s][0815 01:09:46 @base.py:254]  (global_step 146) finished at 1534295386.5130866, time:2.043109178543091.
[0815 01:09:46 @base.py:247]  (global_step 147) begin at 1534295386.5134234
[0815 01:09:47 @base.py:254]  (global_step 147) finished at 1534295387.00269, time:0.48926663398742676.
[0815 01:09:47 @base.py:247]  (global_step 148) begin at 1534295387.0030763
 70%|###################################################################6                            |148/210[03:58<00:46, 1.34it/s][0815 01:09:47 @base.py:254]  (global_step 148) finished at 1534295387.3571424, time:0.3540661334991455.
[0815 01:09:47 @base.py:247]  (global_step 149) begin at 1534295387.3574877
 71%|####################################################################1                           |149/210[04:00<01:39, 0.62it/s][0815 01:09:49 @base.py:254]  (global_step 149) finished at 1534295389.8615568, time:2.5040690898895264.
[0815 01:09:49 @base.py:247]  (global_step 150) begin at 1534295389.8618948
[0815 01:09:50 @base.py:254]  (global_step 150) finished at 1534295390.2188885, time:0.3569936752319336.
[0815 01:09:50 @base.py:247]  (global_step 151) begin at 1534295390.219254
 72%|#####################################################################                           |151/210[04:03<01:25, 0.69it/s]
```

**Comment by yogurfrul** [comment]
I disable warmup by export TF_CUDNN_USE_AUTOTUNE=0 
the problem is still here.
How can multi gpu with allreduce run faster than 1 gpu

**Comment by ppwwyyxx** [comment]
The export will not work because tensorpack enable autotune when import.

I'm sure multi-gpu is not faster than 1 gpu if you wait for like 50 epochs.

**Comment by ppwwyyxx** [comment]
If you still see multi-gpu running faster than 1 gpu, please run it with no modification to the code and post your command and full logs like what's in the issue template.

**Comment by ppwwyyxx** [comment]
I run some tests with today's master code and `--config MODE_MASK=True MODE_FPN=True` on 1 & 8 V100s, with `export TF_CUDNN_USE_AUTOTUNE=0` so the speed is faster and more stable at the beginning.

Starting from the 2nd epoch, 1 V100 takes 2min29s per epoch and 8 V100s take 2min51s per epoch.
